src/backend/sgx/attestation.rs

```diff
@@ -0,0 +1,365 @@
+                ));
+            }
+        }
+
+        let req_len = (buf_wrtr.len() - size_of::<u32>()) as u32;
+        buf_wrtr[0..size_of::<u32>()].copy_from_slice(&req_len.to_le_bytes());
+
+        // Send Request to AESM daemon
+        stream.write_all(&buf_wrtr)?;
+        stream.flush()?;
+
+        // Receive Response
+        let mut res_len_bytes = [0u8; 4];
+        stream.read_exact(&mut res_len_bytes)?;
+        let res_len = u32::from_le_bytes(res_len_bytes);
+
+        let mut res_bytes = vec![0; res_len as usize];
+        stream.read_exact(&mut res_bytes)?;
+
+        Ok(Message::parse_from_bytes(&res_bytes)?)
+    }
+}
+
+/// Gets Att Key ID
+fn get_ak_id() -> Result<Vec<u8>, Error> {
+    let mut stream = UnixStream::connect(AESM_SOCKET)?;
+
+    let r = ReqType::AkIdNum;
+    let mut req = Request::new();
+    let mut msg = Request_GetSupportedAttKeyIDNumRequest::new();
+    msg.set_timeout(AESM_REQUEST_TIMEOUT);
+    req.set_getSupportedAttKeyIDNumReq(msg);
+    let pb_msg: Response = r.send_request(req, &mut stream)?;
+
+    let res = pb_msg.get_getSupportedAttKeyIDNumRes();
+
+    if res.get_errorCode() != 0 {
+        panic!(
+            "Received error code {:?} in GetSupportedAttKeyIDNum",
+            res.get_errorCode()
+        );
+    }
+
+    let num_key_ids = res.get_att_key_id_num();
+
+    let expected_buffer_size: u32 = num_key_ids * SGX_KEY_ID_SIZE;
+    assert_eq!(1, num_key_ids);
+
+    let mut stream = UnixStream::connect(AESM_SOCKET)?;
+
+    let r = ReqType::AkId;
+    let mut req = Request::new();
+    let mut msg = Request_GetSupportedAttKeyIDsRequest::new();
+    msg.set_timeout(AESM_REQUEST_TIMEOUT);
+    msg.set_buf_size(expected_buffer_size);
+    req.set_getSupportedAttKeyIDsReq(msg);
+
+    let pb_msg: Response = r.send_request(req, &mut stream)?;
+
+    let res = pb_msg.get_getSupportedAttKeyIDsRes();
+
+    if res.get_errorCode() != 0 {
+        panic!(
+            "Received error code {:?} in GetSupportedAttKeyIDs",
+            res.get_errorCode()
+        );
+    }
+
+    let key_ids_blob = res.get_att_key_ids();
+    let key_ids: Vec<Vec<u8>> = key_ids_blob
+        .chunks_exact(SGX_KEY_ID_SIZE as usize)
+        .map(Vec::from)
+        .collect();
+
+    assert!(!key_ids.is_empty());
+
+    Ok(key_ids.get(0).unwrap().clone())
+}
+
+/// Fills the Target Info of the QE into the output buffer specified and
+/// returns the number of bytes written.
+fn get_ti(akid: Vec<u8>, size: usize, out_buf: &mut [u8]) -> Result<usize, Error> {
+    assert_eq!(out_buf.len(), SGX_TI_SIZE, "Invalid size of output buffer");
+
+    let mut stream = UnixStream::connect(AESM_SOCKET)?;
+
+    let r = ReqType::TInfo;
+    let pb_req = r.set_request(None, Some(akid), Some(size))?;
+    let mut pb_msg: Response = r.send_request(pb_req, &mut stream)?;
+
+    let res: Response_InitQuoteExResponse = pb_msg.take_initQuoteExRes();
+    let ti = res.get_target_info();
+
+    assert_eq!(
```

Comment:
idem

Reply:
```
    if res.get_errorCode() != 0 {
        return Err(Error::new(
            ErrorKind::Other,
            format!("Received error code {:?} in GetSupportedAttKeyIDNum", res.get_errorCode());
        );
    }
```

Reply:
IMHO this assert is obsolete but in the beginning of the function there is `assert_eq()` that I converted as `debug_assert_eq()`. I'll just remove this one.

Reply:
Also the copy operation is after this check so the error message does not make any sense.